### ISSUE TITLE
Game crashes on loading the save

### ISSUE THREAD

**Issue by Zemogiter** [colobot/colobot#1777]
### Error message

[INFO]: Loading level: levels/free/chapter006/level001/scene.txt

Unhandled exception occurred!
==============================
SIGABRT, abort
==============================

This is usually caused by a bug. Please report this on http://github.com/colobot/colobot/issues
including information on what you were doing before this happened and all the information below.
==============================
You seem to be running a custom compilation of version git-dev~rca4049b8, but please verify that.

The game was in phase PHASE_SIMUL (ID=8)
Last started level was: category=free chap=6 rank=1
==============================

### log.txt

[log.txt](https://github.com/user-attachments/files/17003764/log.txt)

### Save file

[save2408281908VolcanoFree.zip](https://github.com/user-attachments/files/17003771/save2408281908VolcanoFree.zip)

### colobot.ini or colobot.json

[colobot.json](https://github.com/user-attachments/files/17003773/colobot.json)

### Game version (e.g. 0.2.1-alpha)

0.2.1.0+alpha-git-dev~rca4049b8

### Operating system (e.g. Windows 11)

Lubuntu 24.04

### Video card

AMD Radeon R5 M330

### Steps to reproduce the crash

1. Go to "custom game" menu.
2. Press "load".
3. Select any previously saved games.
4. Press "load".
5. Game will crash on "restoring map objects"

### Are you able to reproduce the crsah?

yes

**Comment by hexagonrecursion** [comment]
I have failed to reproduce the issue in vanilla. Based on your colobot.json I see that you have 41 mods installed and 29 of them enabled. It is possible that to reproduce the crash I will need the mods too. Could you please zip up your ~/.local/share/colobot/mods/ and upload it here?

**Comment by Zemogiter** [comment]
There you go https://www.mediafire.com/file/qj0zzg4dgs9sg63/mods.zip/file

**Comment by Zemogiter** [comment]
Tried loading the saves after compiling https://github.com/colobot/colobot/commit/5e63cbf4003b707d74b81c7a055955b4401255ce and I cannot reproduce the error anymore

**Comment by tomangelo2** [comment]
Confirmed on latest dev, while loading save from userlevel
```Unhandled exception occurred!
==============================
Type: CLevelParserExceptionBadParam
Message: Unable to parse 'alludo' as string (param 'dir' in savegame/Gracz/quicksave/data.sav:4)
==============================
```

**Comment by melex750** [comment]
Somehow ```Alludo``` is being saved without double quotes, this seems to fix it...(a different crash happens after that :frowning_face: ):
```diff
--- a/colobot-base/src/level/robotmain.cpp
+++ b/colobot-base/src/level/robotmain.cpp
@@ -4524,7 +4524,7 @@ bool CRobotMain::IOWriteScene(const std::filesystem::path& filename,
     line = std::make_unique<CLevelParserLine>("Mission");
     line->AddParam("base", std::make_unique<CLevelParserParam>(GetLevelCategoryDir(m_levelCategory)));
     if (m_levelCategory == LevelCategory::CustomLevels)
-        line->AddParam("dir", std::make_unique<CLevelParserParam>(GetCustomLevelDir()));
+        line->AddParam("dir", std::make_unique<CLevelParserParam>(StrUtils::ToString(GetCustomLevelDir())));
     else
         line->AddParam("chap", std::make_unique<CLevelParserParam>(m_levelChap));
     line->AddParam("rank", std::make_unique<CLevelParserParam>(m_levelRank));
```

That appears to be the only place where ```CLevelParserParam``` constructor for ```std::filesystem::path``` is invoked.  Maybe that constructor should add double quotes like the one for ```std::string```. @hexagonrecursion ?

**Comment by hexagonrecursion** [comment]
> Confirmed on latest dev, while loading save from userlevel
> 
> ```
> ==============================
> Type: CLevelParserExceptionBadParam
> Message: Unable to parse 'alludo' as string (param 'dir' in savegame/Gracz/quicksave/data.sav:4)
> ==============================
> ```
> 
> Doesn't happen when loading standard levels OS: Kubuntu 23.10


> Somehow `Alludo` is being saved without double quotes, this seems to fix it...(a different crash happens after that ☹️ ):
> 
> ```diff
> --- a/colobot-base/src/level/robotmain.cpp
> +++ b/colobot-base/src/level/robotmain.cpp
> @@ -4524,7 +4524,7 @@ bool CRobotMain::IOWriteScene(const std::filesystem::path& filename,
>      line = std::make_unique<CLevelParserLine>("Mission");
>      line->AddParam("base", std::make_unique<CLevelParserParam>(GetLevelCategoryDir(m_levelCategory)));
>      if (m_levelCategory == LevelCategory::CustomLevels)
> -        line->AddParam("dir", std::make_unique<CLevelParserParam>(GetCustomLevelDir()));
> +        line->AddParam("dir", std::make_unique<CLevelParserParam>(StrUtils::ToString(GetCustomLevelDir())));
>      else
>          line->AddParam("chap", std::make_unique<CLevelParserParam>(m_levelChap));
>      line->AddParam("rank", std::make_unique<CLevelParserParam>(m_levelRank));
> ```
> 
> That appears to be the only place where `CLevelParserParam` constructor for `std::filesystem::path` is invoked. Maybe that constructor should add double quotes like the one for `std::string`. @hexagonrecursion ?

@tomangelo2 @melex750 Your comments appear to be unrelated to this bug report. The save posted by Zemogiter was made from a free game on Volcano - not from a level added by a mod. The colobot.json and mod directory posted above do not even contain any mention of Alludo.

The actual error reported by Zemogiter is caused by this assertion:

https://github.com/colobot/colobot/blob/d8eac1ce6bca423f906d5d9f95f48d4824e6c10b/colobot-base/src/object/object_manager.cpp#L146

Somehow an object with id=88 already exists when we try to create one

**Comment by hexagonrecursion** [comment]
> There you go https://www.mediafire.com/file/qj0zzg4dgs9sg63/mods.zip/file

Saving the zip in case the link goes dead: https://web.archive.org/web/20241126065635/https://download1527.mediafire.com/5d1q9sdnznzg1p9GYdPNSS-k-oDoPHoUAAttOXulFLQlFXsCfrcvtqq17_50LCH9jQhUKaCyQHqDXxE5eEdymuQ13fy7fGEGrDOTOpVriWeXTu6-j1mQ68YxnSyN4CYp8HcowO4URvDNg8f3seN_XvEpxhA61kK9ju51jQyy1QaW/qj0zzg4dgs9sg63/mods.zip

**Comment by hexagonrecursion** [comment]
> Somehow an object with id=88 already exists when we try to create one

1. Somehow the `CreateObject ... id=87 ... type=WheeledGrabber` line in the save file causes `CMotionVehicle::Create()` to be called with `power=0`, which causes a power cell to spawn, which takes the next available id: 88
2. The next CreateObject in the save file also has id=88 causing a collision

**Comment by hexagonrecursion** [comment]
>     1. Somehow the `CreateObject ... id=87 ... type=WheeledGrabber` line in the save file causes `CMotionVehicle::Create()` to be called with `power=0`, which causes a power cell to spawn, which takes the next available id: 88

This is because CObjectManager::ClampPower() replaces -1 with 0: https://github.com/colobot/colobot/blob/d8eac1ce6bca423f906d5d9f95f48d4824e6c10b/colobot-base/src/object/object_manager.cpp#L170-L179

**Comment by hexagonrecursion** [comment]
> Update, [d8eac1c](https://github.com/colobot/colobot/commit/d8eac1ce6bca423f906d5d9f95f48d4824e6c10b) is broken again. Tried to load saves made in [5e63cbf](https://github.com/colobot/colobot/commit/5e63cbf4003b707d74b81c7a055955b4401255ce) and getting the same crash [log.txt](https://github.com/user-attachments/files/17464699/log.txt) Modlist is the same.

Strange. I can reproduce this bug on commit d8eac1ce with no mods at all.

**Comment by hexagonrecursion** [comment]
> I have failed to reproduce the issue in vanilla. Based on your colobot.json I see that you have installed 41 mods and enabled 29 of them. It is possible that to reproduce the crash I will need the mods too. Could you please zip up your ~/.local/share/colobot/mods/ and upload it here?

Strange. I am now able to reproduce the issue the issue in vanilla on ca4049b8 - the same commit I tried before. I must have messed something up last time

**Comment by hexagonrecursion** [comment]
> The actual error reported by Zemogiter is caused by this assertion:
> 
> https://github.com/colobot/colobot/blob/d8eac1ce6bca423f906d5d9f95f48d4824e6c10b/colobot-base/src/object/object_manager.cpp#L146

* The bug was:
    * `power=-1` was being interpreted as `power=0`
    * `CreateObject` with no `power` was also being interpreted as `power=0` (but should have been interpreted as `power=-1`)
* This bug was introduced in #1726 - it appears I somehow forgot to test that the pull request :(

**Comment by hexagonrecursion** [comment]
> > The actual error reported by Zemogiter is caused by this assertion:
> > https://github.com/colobot/colobot/blob/d8eac1ce6bca423f906d5d9f95f48d4824e6c10b/colobot-base/src/object/object_manager.cpp#L146
> 
>     * The bug was:
>       
>       * `power=-1` was being interpreted as `power=0`
>       * `CreateObject` with no `power` was also being interpreted as `power=0` (but should have been interpreted as `power=-1`)
> 
>     * This bug was introduced in [Fix power=0 #1726](https://github.com/colobot/colobot/pull/1726) - it appears I somehow forgot to test that the pull request :(

* Fixed by https://github.com/colobot/colobot/pull/1784

Note: CLevelParserExceptionBadParam is a different bug - I will open a new issue for it

**Comment by hexagonrecursion** [comment]
> The bug was:
> 
>     * `power=-1` was being interpreted as `power=0`
> 
>     * `CreateObject` with no `power` was also being interpreted as `power=0` (but should have been interpreted as `power=-1`)

This is incorrect.

How it should be (_and how it is after_ https://github.com/colobot/colobot/pull/1784):
1. When loading a save, the `power` parameter of `CreateObject` should be ignored - the battery is spawned via CreatePower
    * To achieve this CRobotMain::IOReadObject() overwrites params.power with -1.0f;
    * ClampPower() was replacing -1 with 0 causing the bug
2. When starting a level:
    1. `power=-1` - no battery
        * ClampPower() was replacing -1 with 0 causing the bug
    2. **`CreateObject` without the `power` parameter - a fully charged PowerCell should be spawned**
        * My earlier assertion that `CreateObject` without the `power` parameter should result in no power cell was incorrect
